_posts/2019-08-29-using-tf-from-r/using-tf-from-R.Rmd

```diff
@@ -0,0 +1,413 @@
+#2  0x00007fffe48418f3 in _reticulate_r_to_py_impl (objectSEXP=0x55555ec88fa8, convertSEXP=<optimized out>) at /home/key/R/x86_64-redhat-linux-gnu-library/3.6/Rcpp/include/Rcpp/as.h:151
+...
+#12 0x00007ffff7cc5c03 in dispatchMethod (sxp=0x55555d0cf1a0, dotClass=<optimized out>, cptr=cptr@entry=0x7ffffffeaae0, method=method@entry=0x55555bfe06c0, 
+    generic=0x555557634458 "r_to_py", rho=0x55555d1d98a8, callrho=0x5555555af2d0, defrho=0x555557947430, op=<optimized out>, op=<optimized out>) at objects.c:436
+#13 0x00007ffff7cc5fc7 in Rf_usemethod (generic=0x555557634458 "r_to_py", obj=obj@entry=0x55555ec88fa8, call=call@entry=0x55555c0317b8, args=args@entry=0x55555557cc60, 
+    rho=rho@entry=0x55555d1d98a8, callrho=0x5555555af2d0, defrho=0x555557947430, ans=0x7ffffffe9928) at objects.c:486
+```
+
+
+where `r_to_py` is a generic (like `py_to_r` above), `r_to_py_impl` is wrapped by Rcpp and `r_to_py_cpp` is a C++ function that branches on the type of the object -- basically the counterpart of the C++ `r_to_py`.
+
+Now in addition to that general process, there is more going on when we call an R function from Python. As Python doesn't "speak" R, we need to wrap the R function in CPython - basically, we are extending Python here! How to do this is described in the official [Extending Python Guide](https://docs.python.org/3/extending/index.html#extending-index).
+
+In official terms, what `reticulate` does it _embed_ and _extend_ Python.
+Embed, because it lets you use Python from inside R. Extend, because to enable Python to call back into R it needs to wrap R functions in C, so Python can understand them.
+
+As part of the former, the desired Python is loaded (`Py_Initialize()`); as part of the latter, two functions are defined in a new module named `rpycall`, that will be loaded when Python itself is loaded.
+
+```{Rcpp}
+PyImport_AppendInittab("rpycall", &initializeRPYCall);
+```
+
+These methods are `call_r_function`, used by default, and `call_python_function_on_main_thread`, used in cases where we need to make sure the R function is called on the main thread: 
+
+```{Rcpp}
+PyMethodDef RPYCallMethods[] = {
+  { "call_r_function", (PyCFunction)call_r_function,
+    METH_VARARGS | METH_KEYWORDS, "Call an R function" },
+  { "call_python_function_on_main_thread", (PyCFunction)call_python_function_on_main_thread,
+    METH_VARARGS | METH_KEYWORDS, "Call a Python function on the main thread" },
+  { NULL, NULL, 0, NULL }
+};
+```
+
+`call_python_function_on_main_thread` is especially interesting. The R runtime is single-threaded; while the CPython implementation of Python effectively is as well, due to the [Global Interpreter Lock](https://en.wikipedia.org/wiki/Global_interpreter_lock), this is not automatically the case when other implementations are used, or C is used directly. So `call_python_function_on_main_thread` makes sure that unless we can execute on the main thread, we wait.
```

Comment:
Yeah, I agree that's among the most interesting stuff of everything (at least for me it was ;-)), and I have the paragraphs above the one you cite explaining this:

> Now in addition to that general process, there is more going on when we call an R function from Python. As Python doesn't "speak" R, we need to wrap the R function in CPython - basically, we are extending Python here! How to do this is described in the official [Extending Python Guide](https://docs.python.org/3/extending/index.html#extending-index).

>In official terms, what `reticulate` does it _embed_ and _extend_ Python.
Embed, because it lets you use Python from inside R. Extend, because to enable Python to call back into R it needs to wrap R functions in C, so Python can understand them.

>As part of the former, the desired Python is loaded (`Py_Initialize()`); as part of the latter, two functions are defined in a new module named `rpycall`, that will be loaded when Python itself is loaded.

```{Rcpp}
PyImport_AppendInittab("rpycall", &initializeRPYCall);
```
>These methods are `call_r_function`, used by default, and `call_python_function_on_main_thread`, used in cases where we need to make sure the R function is called on the main thread: 

```{Rcpp}
PyMethodDef RPYCallMethods[] = {
  { "call_r_function", (PyCFunction)call_r_function,
    METH_VARARGS | METH_KEYWORDS, "Call an R function" },
  { "call_python_function_on_main_thread", (PyCFunction)call_python_function_on_main_thread,
    METH_VARARGS | METH_KEYWORDS, "Call a Python function on the main thread" },
  { NULL, NULL, 0, NULL }
};
```

The reason I'm talking about `call_python_function_on_main_thread` is a bit special: In `keras`, there are occasions we need to pay attention we're actually doing that (calling Python on the main thread). I didn't want to go too much into examples though, for several reasons (one of them being, this seemed a bit anticlimactic)